src/main/java/net/datafaker/Unique.java

```diff
@@ -0,0 +1,60 @@
+package net.datafaker;
+
+import java.util.*;
+
+public class Unique extends AbstractProvider {
+
+    private final Map<Locale, Map<String, List<String>>> valuesByKeyAndLocale = new HashMap<>();
+
+    public Unique(Faker faker) {
+        super(faker);
+    }
+
+    public String fetchFromYaml(String key) {
+        Locale locale = faker.getLocale();
+
+        Map<String, List<String>> valuesByKey = valuesByKeyAndLocale.getOrDefault(locale, new HashMap<>());
+        List<String> values = valuesByKey.get(key);
+
+        if (values == null) {
+            values = fetchValues(key);
+        }
+
+        if (values.isEmpty()) {
+            throw new RuntimeException(String.format(
+                "All possible values have been generated for key %s under locale %s",
+                key,
+                locale
+            ));
+        }
+
+        int index = faker.random().nextInt(0, values.size() - 1);
+        String value = values.remove(index);
```

Comment:
To speed up removal of elements it is better to do in 2 steps:
1. exchange element with index `index` with the latest element
2. remove the latest element.
This will be much faster (O(1) complexity vs current O(n)) just because in that case there is no need to move elements left.
The downside  is no ordering however in this case we don't need any order

Reply:
Done. Thanks for the advice.